doc/decisions/0_drafts/errors_core.md

```diff
@@ -0,0 +1,403 @@
+
+This leaves the API design entirely open.
+
+TLS was only standardized in C11, while Elektra currently only requires C99.
+
+This has all the same issues as `errno`: needs to be cleared to detect errors, can only store one error (per thread), etc.
+
+### Limit to single error value
+
+Always limiting a function to a single error code can be very limiting and in some cases and makes it very hard to find out what the exact problem was.
+Reducing a function to a single error value, makes sense if the errors are easy to distinguish.
+But for more complex preconditions, this can be annoying and inefficient to use.
+For example, a function `foo` that takes two keynames, both of which must be valid.
+To find out which of the names was invalid, the caller would have to call something like `elektraCheckName(name)` on at least one name.
+That's a non-trivial check that is not cheap in terms of runtime.
+More importantly, it is a check that `foo` already did, and the result could just be propagated to the caller.
+
+### Redesign API
+
+Similar to the option above, we limit all functions to a single error value.
+But we also redesign the API, so that every API function only has maximal one error case (apart from failed allocation and `NULL` pointers as arguments).
+
+- `nameNew(name) -> name_object` only has the error case of invalid name
+- `keyNew(name_object)` won't have error cases anymore
+- `keySetName(name_object)` only has the error case of read-only names
+
+The advantage is that an API like this is hard to be used incorrectly.
+You can only call `keySetName` when you already have a valid key name, so there is only a single error case for `keySetName`.
+If `keySetName` fails, we always know exactly what was wrong; the name was read-only.
+
+However, always partitioning an API like this could create very annoying to use APIs.
+It could also have a negative impact on memory use.
+To make an API like this work properly, all the intermediate objects that allow us to have functions with a single error case, must be opaque structs.
+That means they must be heap-allocated as separate objects.
+
+### Require caller to check preconditions
+
+If caller are interested in the specific errors, they need to do extra checks, e.g., before calling `keySetName`, they call `elektraCheckName(name)` or `keyIsLocked`.
+
+This is not so intuitive and is against our principle to make it hard to use the API wrong.
+
+It's also unclear, what the function would do, if the caller did not check the preconditions.
+
+### Pragmatic solution
+
+We assume the functions in question are side-effect-free and the only possible cause of errors are invalid arguments.
+We also assume that functions _can_ always be written such that they fail before any permanent changes to the arguments are made.
+
+If we assume that _all_ functions are indeed written that way, then we can conclude, that in theory a single error indicating return value is always enough.
```

Comment:
No assumption needed, it would be the new guideline that functions must be that way. (And reviewed that way, see #4426)

Reply:
If you read on, this solution actually argues **against** introducing such a strict rule, because it could introduce unwanted restrictions.

I'll give you an example that shows how tedious such a rule could become. (I'll add something like this to the file as well).

Let's first assume there is a function `foo()` that needs a valid escaped keyname as an argument. There are multiple reasons why an escaped keyname could be invalid. The user may want to differentiate between those reasons. With the strict "1 error value" rule, the result of `foo()` can't tell us what the issue with the keyname is. But that's fine, we'll just add a `checkName()` function. However, that still must follow the "1 error value" rule. So what we would actually need is `checkNameProblem1()`,  `checkNameProblem2()`, etc. each of which just checks 1 potential problem and the user has to call all of them.

Of course, I assume, that is not what you actually want. So I once again think we are on the same page here. The rule should be: "Try to use only 1 error value, but a useful API is more important" That is more or less, what this solution argues for, just in lot's more detail, because I wrote down all the cases I thought about.

> **Note**: I know the example is somewhat abstract and doesn't use examples from the API. But since we are talking about potential rules for unknown future API, we should think abstractly and not limit ourselves to only the cases we have right now.